File: golang/problems/no_2/no_2.go
package no_2

import (
	"awesomeProject/golang/utils/linkedList"
)

//2. Add Two Numbers
//You are given two non-empty linked lists representing two non-negative integers. The digits are stored in reverse order, and each of their nodes contains a single digit. Add the two numbers and return the sum as a linked list.
//
//You may assume the two numbers do not contain any leading zero, except the number 0 itself.

//type ListNode struct {
//	Val  int
//	Next *ListNode
//}

func addTwoNumbers(l1 *linkedList.Node, l2 *linkedList.Node) *linkedList.Node {
	//total := 0
	carry := 0

	ret := linkedList.Node{
		Val:  -1,
		Next: nil,
	}
	p := &ret
	// handle common part
	for l1 != nil && l2 != nil {
		curSum := l1.Val + l2.Val + carry
		carry = curSum / 10
		curSum %= 10
		l1.Val = curSum
		p.Next = l1
		p = p.Next
		l1 = l1.Next
		l2 = l2.Next
		//total += curSum * math.Pow(10, digit)
	}
	// handle the rest
	for l1 != nil {
		curSum := l1.Val + carry
		carry = curSum / 10
		curSum %= 10
		l1.Val = curSum
		p.Next = l1
		l1 = l1.Next
		p = p.Next
	}

	for l2 != nil {
		curSum := l2.Val + carry
		carry = curSum / 10
		curSum %= 10
		l2.Val = curSum
		p.Next = l2
		l2 = l2.Next
		p = p.Next
	}

	if carry != 0 {
		p.Next = &linkedList.Node{
			Val:  carry,
			Next: nil,
		}
	}

	return ret.Next
}

File: python/no_2/no_2.py
class Solution:
    def addTwoNumbers(self, l1: Optional[ListNode], l2: Optional[ListNode]) -> Optional[ListNode]:
        flag = 0

        v1, v2 = ListNode(-1, l1), ListNode(-1, l2)
        pp1, pp2 = v1, v2
        p1, p2 = l1, l2
        while p1 and p2:
            # print(v1)
            cur_sum = p1.val + p2.val + flag
            p1.val = cur_sum % 10
            flag = cur_sum // 10

            pp1 = p1
            pp2 = p2
            p1 = p1.next
            p2 = p2.next

        if p1 is None:
            pp1.next = p2
            while p2:
                cur_sum = p2.val + flag
                p2.val = cur_sum % 10
                flag = cur_sum // 10

                pp1 = p2
                p2 = p2.next
        elif p2 is None:
            while p1:
                cur_sum = p1.val + flag
                p1.val = cur_sum % 10
                flag = cur_sum // 10

                pp1 = p1
                p1 = p1.next

        if flag:
            pp1.next = ListNode(flag)

        return v1.next